Armclang / armlink support in Zephyr - Zephyr CMake linker functions

tejlmand opened the pull request:
This is the initial support for Zephyr CMake linker functions.

Using CMake for linker script generations allows us to specify memory regions, linker sections, linker symbols in a common way using CMake functions, and then generate toolchain specific linker scripts / scatter files.

This allows users to only write things once and avoid linker scripts to get out of sync when only a single script (for example ld script) is updated, but not other linker scripts.

Besides a unifies Linker script language, another advantage of using CMake is that users can add new sections in their application CMakeLists.txt file.

Using CMake, a user can now write:
```
zephyr_linker_memory(NAME FOO FLAGS rx START 0x40000000 SIZE 1k)
zephyr_linker_section(NAME bar LMA FOO)
zephyr_linker_section_configure(SECTION bar INPUT baz KEEP)
```

and this can even be done in the app CMakeLists.txt file, like:
```
find_package(Zephyr REQUIRED)

target_sources(app PRIVATE baz.c)

zephyr_linker_memory(NAME FOO FLAGS rx START 0x40000000 SIZE 1k)
zephyr_linker_section(NAME bar LMA FOO)
zephyr_linker_section_configure(SECTION bar INPUT baz KEEP)
```

This will create a memory region named `FOO`, with a section `bar` where the input section `baz` will be placed with the KEEP flag.

Also it is possible to describe a new section in CMake together with the library as:
```
zephyr_library_named(pinmux)
zephyr_library_sources(pinmux.c)
zephyr_linker_section(NAME .pinmux GROUP ROM_REGION)
or
zephyr_linker_section(NAME .pinmux LMA ROM)
```

With this PR, the arm architecture has been ported to the new Zephyr CMake linker generator.

When building with armclang, the new linker generator is always used.
When building with `zephyr` or `gnuarmemb`, the old system will be used, and the new linker generator must be manually selected using: ` -DCONFIG_CMAKE_LINKER_GENERATOR=y` (or by using menuconfig).

To build using `armclang` set the toolchain to armclang and the path to the toolchain.
```
export ZEPHYR_TOOLCHAIN_VARIANT=armclang
export ARMCLANG_TOOLCHAIN_PATH=/opt/armds-2020-1/sw/ARMCompiler6.15/
```

Building `hello_world` for `qemu_cortex_m3` using armclang / armlink:
```
$ cmake -GNinja -DBOARD=qemu_cortex_m3 -Bbuild_qemu samples/hello_world/
...
-- The C compiler identification is ARMClang 6.15.2
-- The CXX compiler identification is ARMClang 6.15.2
-- The ASM compiler identification is ARMClang
...
$ ninja -Cbuild_qemu run
[116/117] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: cortex-m3
qemu-system-arm: warning: nic stellaris_enet.0 has no peer
*** Booting Zephyr OS build zephyr-v2.6.0-69-gcfa5059ceaeb  ***
Hello World! qemu_cortex_m3
```

and for `nrf52840dk_nrf52840`:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/hello_world/
...
$ ninja -Cbuild run
$ ninja -Cbuild flash
...
...
-- runners.nrfjprog: Board with serial number 683845540 flashed successfully.
```
Open minicom and see the console output:
```
*** Booting Zephyr OS build zephyr-v2.6.0-69-gcfa5059ceaeb  ***
Hello World! nrf52840dk_nrf52840

```

Building with Zephyr toolchain using new linker generator:
```
$ export ZEPHYR_TOOLCHAIN_VARIANT=zephyr
$ cmake -GNinja -DBOARD=qemu_cortex_m3 -Bbuild_qemu samples/hello_world/ 
 -DCONFIG_CMAKE_LINKER_GENERATOR=y
....
$ ninja -Cbuild_qemu run
[124/125] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: cortex-m3
qemu-system-arm: warning: nic stellaris_enet.0 has no peer
Timer with period zero, disabling
*** Booting Zephyr OS build zephyr-v2.6.0-69-gcfa5059ceaeb  ***
Hello World! qemu_cortex_m3
```

ToDo:
- [ ] Documentation
- [x] Kconfig setting for linker script generator
- [x] Memory sections from Device Tree
- [x] Linker groups / regions
- [ ] Userspace support
- [x] Linker symbols for groups
- [x] offset on section configure

Additional notes:
- CMake v3.18 or higher required, see #37575
- ~Issues seen when using `make`, so for time being, please use `ninja`~ CMake 3.21 with make is supported. Older CMake with make will issue a error message, see: https://github.com/zephyrproject-rtos/zephyr/commit/d72ec887cdc51a8b5958756888b69f608fd1245b
- Warnings will be seen during linking. Those warnings will be cleaned up before going into mergable PR state.
- In general, memory regions should be defined using devicetree, and then Zephyr CMake linker template will create corresponding memory regions in the linker script, see #37624

The cleanup commit series can be found here as independent PR:
https://github.com/zephyrproject-rtos/zephyr/pull/37938

Fixes #33435

tejlmand committed:
armclang: support for armclang compiler and armlink linker

This is the initial support for the armclang compiler together with the
armlink linker.

Introduced in this commit:
- armclang compiler support
- armlink linker support
- armlink scatter file generator for scatter loading
- dual pass linker script generation

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
asm: .eabi_attribute Tag_ABI_align_preserved, 1

Tell armlink that files has ensured proper stack alignment.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
armclang: adding toolchain/armclang.h header

The armclang.h includes the toolchain/llvm.h header.

Also it redifines the __GENERIC_SECTION and Z_GENERIC_SECTION so that
they includes the `used` attribute which is needed by armlink.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: linker: converter arm and common ld scripts into CMake code

Converted existing ld script templates into CMake files.

This commit takes the common-ram.ld, common-rom.ld, debug-sections.ld,
and thread-local-storage.ld and creates corresponding CMake files for
the linker script generator.

The CMake files uses the new Zephyr CMake functions:
- zephyr_linker_section()
- zephyr_linker_section_configure()
- zephyr_linker_section_obj_level()

to generate the same linker result as the existing C preprocessor based
scheme.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
armclang: bintools support for armclang (arm-ds/Keil)

Initial bintools support for elfconvert
- hex, srec, bin conversion
- strip
- gap fill

Initial bintools support for readelf
Initial bintools support for disaasembly

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
armclang: specify armlink link executable signature

The default armlink signature uses `--list=<TARGET_BASE>.map`, but in
Zephyr we uses a different name for the map file, therefore we need to
specify a custom link executable signature.

This is done in the linker specific flags file:
cmake/linker/armlink/linker_flags.cmake

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
scripts: gen_handles.py: take device start symbol as argument.

The current gen_handles.py script uses linker defined symbols.
As preparation for support of more linkers the gen_tables.py now takes
the device start symbol as argument.

For example, armlink and ld uses different symbols.
With ld those can be named explicitly, but for armlink the linker
decides the names.

For ld, Zephyr defines: __device_start
For armlink, the symbol is defined as: Image$$<section name>$$Base

Therefore knowledge of the linker symbol to be used must be passed to
gen_handles.py so that the correct symbol can be used for locating
devices.

To support this change, the creation of the asm, compiler, compiler-cpp,
linker targets has been moved from target_toolchain_flags.cmake to
target_toolchain.cmake.

All linkers has been updated to support the use of the
device_start_symbol on the linker target.

List of linkers updated:
- ld
- lld
- arcmwdt

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
arch: linker: specify intList section in the IDT_LIST region

This commit specifies the intList section in the IDT_LIST region in the
arch/common CMakeLists.txt file.

It uses zephyr_linker_section to setup the intList section for first
pass linker file and configures the section to hold irq_info and
intList input section.

For second pass linker file, the irq_info and intList input sections are
placed in the /DISCARD/ section.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
armclang: armlinker: armlink symbol steering file

This commit introduces armlink steering file.

A steering file in armlink allows Zephyr to keep using existing linker
symbols defined in ld scripts and used throughout the code tree.

The steering file is generated at build time in order to resolve Zephyr
linker symbols to their corresponding armlink symbols.

As example, Zephyr defines __ramfunc_start which corresponds to the
armlink auto defined Image$$ramfunc$$Base symbol.

Or __init_PRE_KERNEL_1_start which corresponds to Image$$init_0$$Base.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
kconfig: cmake: CMake linker script generator symbol added

A "Linker script" choice group has been added to the linker options
menu.

The existing HAVE_CUSTOM_LINKER_SCRIPT config has been relocated inside
the linker script menu.

Two new entries has been created:
- LD template, for using the old ld template files together with the C
  pre-processor to create the final linker script.
- CMake linker generator, used for ARMClang / armlink toolchain.
  The CMake linker generator is also supported for LD linker scripts
  for the ARM architecture.

In CMake, the file cmake/linker/ld/target.cmake has been updated to
support the CMake LD linker generator.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: CMake linker functions implemented

This commit introduces zephyr_linker CMake functions for creation of
linker scripts.

The linker functions supports the following features:
- Configuration of memory sections
- Configuration of memory groups
- Creation of output sections
- Configuration of input section the should go into output section and
  how those should be treated, such as sorting, keep, priority.
- Defining linker symbols
- Specifying Kernel VMA to support virtual linking on x86

Overview of functions and macros introduce with this commit:
- pow2round
- zephyr_linker
- zephyr_linker_memory
- zephyr_linker_memory_ifdef
- zephyr_linker_group
- zephyr_linker_section
- zephyr_linker_section_ifdef
- zephyr_iterable_section
- zephyr_linker_section_obj_level
- zephyr_linker_section_configure
- zephyr_linker_symbol
- zephyr_linker_property_append

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: ld: adding ld linker script generator

Adding intial version of ld_script.cmake.

This script can generate ld linker script from the Zephyr CMake linker
functions.

This script is called by the build system and all linker settings, such
as memory regions, groups, output sections, linker symbols defined with
Zephyr CMake linker functions are passed to this script in order to
generate an ld linker script.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: linker: arm: adding Zephyr CMake linker files for arm arch

This is the initial version of a Zephyr CMake linker file for the arm
architecture.

This file defines memory regions, groups, linker sections and symbols
for the arm architecture.

It also sources the common common-ram.cmake, common-rom.cmake,
debug-sections,cmake, and thread-local-storage.cmake.

It configure sections for SoC families using zephyr_linker_sources()
functions:
- nxp_imx

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

galak reviewed (commented):
I think some of these commits we could pull out and merge sooner:
*  cmake: removed stray space in readelf outfile flag 
*  linker: removing unused _DATA_IN_ROM 
*  linker: align __data_ram/rom_start/end linker symbol names 
etc.

galak commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
   )
 endfunction()
+
+########################################################
```
I think we should put these in their own file (cmake/linker.cmake) and think about removing the `zephyr_` prefix as I think this could be used by other projects.

tejlmand replied:
makes sense.

tejlmand replied:
@galak, not sure if `linker.cmake` is best choice directly at the `<zephyr>/cmake/` level as it may not be obvious to people that this particular file contains CMake extension functions / macros.

Maybe we should refactor and the create:
```

tejlmand commented:
> 
> 
> I think some of these commits we could pull out and merge sooner:

I completely agree, and this was also my intention, but I wanted to create this PR first so people could se the reasoning for such independent cleanup.

tejlmand commented on arch/arm/core/aarch32/cortex_m/vector_table.S:
```diff
@@ -22,6 +22,10 @@
 #include "vector_table.h"
 
 _ASM_FILE_PROLOGUE
+/*
+ * Tell armclang that stack alignment are ensured.
+ */
+.eabi_attribute Tag_ABI_align_preserved, 1
```
very good point, thanks

tejlmand commented on lib/os/cbprintf_packaged.c:
```diff
@@ -32,13 +32,13 @@ static inline bool ptr_in_rodata(const char *addr)
-	extern char _image_rodata_end[];
-#define RO_START _image_rodata_start
-#define RO_END _image_rodata_end
+	extern char __rom_start_start[];
```
> I'm confused about this patch. Let's take this line as an example.
> 
> This line changes RO_START from _image_rodata_start to _rom_start_start on most architectures.

I think you're right at multiple things.
First, the part of aligning linker symbol is still at first stage, as I have focused on the CMake part to provide enough of the CMake linker feature to see that this direction is indeed feasible, and as part of that I have only done the necessary cleanup to get `hello world` running when using both existing infrastructure, but also with generated linker scripts (ld and scatter).

In scatter files we cannot control symbol ourselves, instead armlink will provide linker defined symbols, based on section names. This is why a steering file is used to link the names we uses to names provided by armlink.
In general I prefer that we (Zephyr) defines the names we want to use, and not the linker we use.

Most places, Zephyr also define linker symbols around output sections which almost follows the name of the section.
The `__image_rom_X` and `__image_rodata_X` being exceptions.
Now, handling of such groups are still wip, and can indirectly be seen by this commit message: https://github.com/zephyrproject-rtos/zephyr/pull/36140/commits/0dd656f18b65072d0562786e4a3b40aaea76d478 but I should probably have made that clearer in the PR description.

That being said, in this particular case, the switch from `_image_rodata_start` to `__rom_start_start` does seem to be wrong.
So far, there is no correct replacement ready for the `_image_rodata_X` symbols.
(there is also no real replacement for `_image_rom_start` and `_image_rom_end`, but for `_image_rom_start` we do have a `rom_start` section that are close, but it only matches the start symbol for the rom_start section, not the end symbol).

With regards to bisectability, the plan is that all cleanup needed to prepare for CMake linker feature should comes first, so that they can be taken in independently, as commented here: https://github.com/zephyrproject-rtos/zephyr/pull/36140#issuecomment-858781685 but i'm not there yet.

Finally, thanks for the nice observation and I will surely take a closer look at `_image_rodata_X`.
Also I hope the additional description gives a little better understanding of the direction and progress of the work.

mbolivar-nordic replied:
That's helpful, thanks. By the way, will this abstraction with scatter and steering files make it possible to resolve this linker issue which shows up with GNU ld on some out of tree architectures?

https://lists.zephyrproject.org/g/devel/message/7765

mbolivar-nordic replied:
That's helpful, thanks. By the way, will this abstraction with scatter and steering files make it possible to resolve this linker issue which shows up with GNU ld on some out of tree architectures?

https://lists.zephyrproject.org/g/devel/message/7765

mbolivar-nordic replied:
Just to be clear, I know that GNU ld and armlink are not the same program. I'm asking if the abstraction you're adding within the build system that lives on top of these two linkers could be adapted to solve the problem of GNU ld also changing the names of some symbols in the case of this Renesas system and apparently some others.

tejlmand replied:
@mbolivar-nordic I do think the linker script generator will offer some possibilities to address the issue regarding symbols and the Renesas system.

The exact way of doing so will of course require some exploration work, but from the description in the link I think the linker abstraction has a high probability of being able to solve such an issue.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+# Zephyr linker general settings.
+# This function specifies general settings for the linker script to be generated.
+#
+# FORMAT <format>: The output format of the linked executable.
```
The function `zephyr_linker()` itself should support being called more than once, but with different arguments.

Currently it only handles output format, but another use-case could be a default alignment.
Today, we do use `_region_min_align` but at current stage, i have not yet decided myself how this is to be handled in the new scheme, but setting this through `zephyr_linker` could be an option, and that setting could be set independently of `FORMAT`.

But calling `zephyr_linker(FORMAT <val>)` twice should be an error, so that will be fixed.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+# This function specifies a memory region for the platform in use.
+#
+# Note:
+#   This function should generally be called with values obtained from device
+#   tree or Kconfig.general settings for the linker script to be generated.
```
> "devicetree" is preferred to "device tree"

Fixed.

> I can't parse "or Kconfig.general settings for the linker script to be generated."

Me neither, so following sentence removed.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+#                  r: Read-only region
+#                  w: Read-write region
+#                  x: Executable region
+#                  The flags r and x, or w and x may be combined like: rx, wx.
```
Good point.
For a start I will leave it a bit open, to allow some more thoughts on best approach here.

@TTornblom would there be something to consider for IAR in this regard ?

TTornblom replied:
@tejlmand I will check this out, thanks.

mbolivar-nordic replied:
> I will check this out, thanks.

@TTornblom ping

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+endfunction()
+
+# Internal helper macro for zephyr_linker*() functions.
+macro(zephyr_linker_property_append list arguments)
```
No, it's appending to the list provided and that list goes to the linker property.
But I see how the name can cause confusion, so will look into a better name later.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+    )
+  endif()
+
+  zephyr_linker_property_append("MEMORY" "${single_args}")
```
It depends on where we actually want to control this part.

If we obtain the knowledge of those memory regions from devicetree or Kconfig, should this be ensured there ?
and could we then get into that situation at all.

Or is the CMake part here just passing this on to the script that actually generates the linker, so that script decides if the field is valid or not.

For now, I have tried to minimize the logic in the linker functions in CMake on what is valid values, but exactly on the memory it might make sense to do some basic checking, as it allows us to provide good error messages at CMake configure time, and thus make debugging earlier.

mbolivar-nordic replied:
> If we obtain the knowledge of those memory regions from devicetree or Kconfig, should this be ensured there ?
> and could we then get into that situation at all.

My opinion: the variable name MEMORY is pretty generic and we shouldn't assume it's not set.

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+  if(LINKER_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "zephyr_linker_memory(${ARGV0} ...) given unknown "
+                        "arguments: ${MEMORY_UNPARSED_ARGUMENTS}"
+    )
```
Fixed

tejlmand commented on kernel/main_weak.c:
```diff
@@ -0,0 +1,14 @@
+
+#include <kernel_internal.h>
+
+void __weak main(void)
```
done

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2245,3 +2279,379 @@ function(target_byproducts)
+# same name, unless NOINPUT is specified.
+# This means the section named `foo` will default include the sections matching
+# `foo` and `foo.*`
+# Each output section will define the followiing linker symbols:
```
fixed.

dleach02 commented:
Addresses #33435

cfriedt commented:
@tejlmand 

I wonder if this might tie in with #37123 (macos native_posix support).

The major issue with macos is that the Mach-O binary file format only supports section names 1 <= n <= 16 chars in length, and we generate a number of linker-defined sections that exceed that size.

stephanosio commented:
> I wonder if this might tie in with #37123 (macos native_posix support).

@cfriedt I don't think so. This PR has nothing to do with Mach-O format or even Clang in general.

armclang is not Clang. It refers to the commercial Arm Compiler (from Arm) that is based on Clang but so heavily modified that it is not really the Clang as you know it.

mbolivar-nordic commented:
> @cfriedt I don't think so. This PR has nothing to do with Mach-O format or even Clang in general.

@stephanosio for context, I suggested that @cfriedt look at this PR because of commits like https://github.com/zephyrproject-rtos/zephyr/pull/36140/commits/5bf4e11e84650641eb6349e0e79491f33af2d76a. The problem on Mach-O is that section names can only be 16 characters long, so many existing zephyr linker script sections used for ELF which are longer than that won't fit. The abstraction layer between 'zephyr' symbols and 'linker' symbols being added here seemed potentially relevant.

tejlmand commented:
> The major issue with macos is that the Mach-O binary file format only supports section names 1 <= n <= 16 chars in length, and we generate a number of linker-defined sections that exceed that size.
> 
> One possible solution for that would be some kind of mapping for section names. I'll be looking at other solutions as well. Currently macos support is very brute force.

@cfriedt Thanks for poiting my attention to this issue.
This for sure is something we should be able to handle in the linker script generator.
In code we rely on the linker script defined symbols, so we should be able to truncated the section names to <16 when building on macos, for native posix.

Will keep that in mind and probably ask for your review when this PR is out of draft, to get feedback on the macos part :wink:

cfriedt commented:
> @cfriedt Thanks for pointing my attention to this issue.
> Will keep that in mind and probably ask for your review when this PR is out of draft, to get feedback on the macos part 😉

@tejlmand - I would be happy to test out anything for you or even listen if you need to bounce some ideas around 😁

tejlmand committed:
cmake: ld: adding common linker script generator

Adding intial version of linker_script_common.cmake.

This script parses memory regions, groups, sections, settings, and
symbols defined in the Zephyr CMake build system.

The linker script creates objects for each type above and groups
sections, symbols, and groups together based on their configuration.

This creates a hierarchy which the `<linker>_script.cmake`
implementations can then use during linker script generation.

linker_script_common.cmake also provides stubs for <type>_to_string()
functions which are used when generating the output script.

Each specific `<linker>_script.cmake` must implement those to_string()
functions to create the final linker script.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
An update has been posted.

This update contains:
- complete rewrite of scatter_script.cmake
- complete rewrite of ld_script.cmake
- Introduction of linker_script_common.cmake which are shared among linker script generators
- group feature added which means that multiple sections can be added to a group to ensure they are placed together.
  - This includes linker symbols defined for the start and end of such groups, `__<group>_start` and `__<group>_end`, for example `__rom_region_start` and `__rom_region_end` will be defined for the group `rom_region`.
  - Groups can be placed inside another group, for example:
    ```
    zephyr_linker_group(NAME ROM_REGION LMA FLASH)
    zephyr_linker_group(NAME TEXT_REGION GROUP ROM_REGION)
    ```
    resulting in symbols and location like:
    ```
    __rom_region_start;
    __text_region_start;
    __text_region_end;
    __rom_region_end;
    ```
- more cleanup 

Also addressed:
> But common-rom.ld doesn't define a __rom_start_start symbol at the beginning, so this looks like it's now pointing to the wrong thing. 

With group feature in place, this should now be correct.

See diff, before rebase: https://github.com/zephyrproject-rtos/zephyr/compare/3191ee9dec1884a874199fb682eace4125bf3082..4a5f028599aeec1647a12a8591bc6262927ed2fd

tejlmand committed:
cmake: CMake devicetree related linker functions implemented

This commit introduces zephyr_linker_dts CMake functions for creation of
linker scripts based on devicetree nodes.

The linker devicetree functions supports the following features:
- Configuration of memory sections based on devicetree nodes

Overview of functions introduced with this commit:
- zephyr_linker_dts_memory

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: linker: arm: memory regions from devicetree added

This commit add devicetree memory regions added to arm/linker.cmake.

The following memory regions are now added to the generated linker
script and scatter file if they exists in the devicetree.
- chosen:    zephyr,itcm
- chosen:    zephyr,dtcm
- nodelabel: ti_ccfg_partition
- nodelabel: sram1
- nodelabel: sram2
- nodelabel: sram3
- nodelabel: sram4
- nodelabel: sdram1
- nodelabel: sdram2
- nodelabel: backup_sram

Also support for the itcm and dtcm section and their placement in
corresponding memory regions has been added.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
armclang: warn on non-Ninja generator for CMake 3.20.

When using ARMClang linker and scatter files (armlink) then all
libraries are linked as object libraries using `$<TARGET_OBJECTS:lib>`.

CMake version 3.20 only has limited support for such linking:
> Referencing $<TARGET_OBJECTS> in target_link_libraries calls worked
> in versions of CMake prior to 3.21 for some cases, but was not fully
> supported.

One of those cases that do not work is Unix Makefiles generators.

As only Ninja is currently verified to work, this commit will check the
CMake version in use and the generator, and if CMake version <=3.21 is
used with non-Ninja generator then an error is raised informing user
to either use Ninja generator or update CMake.

As the Ninja generator has been confirmed to work as expected, then that
generator is accepted with CMake 3.20 and older.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
The ARMClang / scatter file is now considered ready for review.

tejlmand committed:
armclang: ARM Compiler C library support

Support for ARM Compiler C library.

This commit add support for the ARM Compiler C libary in:
- Kconfig
- libc/armstdc

A new Kconfig symbol is added to allow a toolchain to specify if they
support linking with the minimal C library.
Also the CMake variable `TOOLCHAIN_HAS_NEWLIB` is exported to Kconfig
so that CONFIG_NEWLIB_LIBS can only be enabled if the toolchain has
newlib.

The armclang toolchain selects the CMake scatter file generator and
disables support for the LD linker template which is not supported by
armlink.

For the ARM Compiler C library, a corresponding lib/libc/armstc/ folder
with a minimal implementation to work with the ARM Compiler C library
is added.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
Some testing results:
Building
```
+---------------------+----------------+----------------+------------------+
|          -          |  hello world   |     blinky     | bt peripheral hr |
+---------------------+----------------+----------------+------------------+
| Board               | ld  |  scatter | ld  |  scatter | ld  |  scatter   |
| qemu_cortex_m3      |  X  |    X     |     |          |     |            |
| nRF51dk_nrf51422    |  X  |    X     |  X  |    X     |  X  |    X       |
| nRF52840dk_nrf52840 |  X  |    X     |  X  |    X     |  X  |    X       |
| mimxrt1020_evk      |  X  |    %     |  X  |    %     |  X  |    %       |
+---------------------+----------------+----------------+------------------+
```
Note, the mimxrt1020_evk fails to link because the hal_nxp/imx takes a different codepath when compiling with ARMClang, causing it to be dependent DefaultISR symbol. This code must be adopted to support ARMClang / Zephyr.
https://github.com/zephyrproject-rtos/hal_nxp/blob/3b45fbaae29640ced4eafdd9e63fe534ca1401ca/mcux/drivers/imx/fsl_lpuart.c#L96

Running:
```
+---------------------+----------------+----------------+------------------+
|          -          |  hello world   |     blinky     | bt peripheral hr |
+---------------------+----------------+----------------+------------------+
| Board               | ld  |  scatter | ld  |  scatter | ld  |  scatter   |
| qemu_cortex_m3      |  X  |    X     |     |          |     |            |
| nRF51dk_nrf51422    |  X  |    X     |  X  |    X     |  %  |    %       |
| nRF52840dk_nrf52840 |  X  |    X     |  X  |    X     |  %  |    %       |
+---------------------+----------------+----------------+------------------+
```
The runtime failures for bluetooth is being investigated.

carlescufi commented:
I would like to discuss the steps required to merge this into main, along with the maintenance associated with it if `CONFIG_CMAKE_LINKER_GENERATOR` is set to `n` by default. This has nothing to do with armclang itself, it's about switching to the new CMake-based linker script generation even for GCC.

dleach02 commented:
The referenced problem with NXP HAL is because we don't have the DefaultISR defined anywhere. With the NXP SDK, this is a "weak" symbol that is defined in our startup code. This was added to our codebase a while back to deal with a code size issue. Apparently, the armclang tool chain has a quirky behavior with pointers like this that eventually get set within a function. If that function is not referenced by the rest of the code, tool chains like GCC will properly prune that function from the final image. The armclang though does not. Something to do with a pointer that is initially pointing to a NULL.

galak commented:
dev-review (Aug 19, 2021):
* linker templates needed for architectures (only arm supported today) `ARM: cmake/linker_script/arm/linker.cmake `
- merge w/default disable, add a tests to make sure it doesn't bitrot
- look at possibly moving cmake/linker_script/arm/linker.cmake into arch dir instead

* arm-clang enablement:
- need a broader plan
- look at implementing a "quick" test
- test on qemu_cortex_m3

* which libc are we getting?  (needs some verification, minlibc v arm-libc)

mbolivar-nordic commented on arch/common/ramfunc.ld:
```diff
@@ -9,12 +9,12 @@
 	*(".ramfunc.*")
-	MPU_ALIGN(_ramfunc_ram_size);
-	_ramfunc_ram_end = .;
+	MPU_ALIGN(__ramfunc_size);
```
I know this exists before this PR, but can you really read `__ramfunc_size` here before setting it on line 19 below?

mbolivar-nordic commented on arch/arm64/core/mmu.c:
```diff
@@ -651,8 +651,8 @@ static const struct arm_mmu_flat_range mmu_zephyr_ranges[] = {
 	{ .name  = "zephyr_rodata",
-	  .start = _image_rodata_start,
-	  .end   = _image_rodata_end,
+	  .start = __rodata_region_start,
```
This patch is missing some renames too.

```
$ rg '_image_rom_start|_image_rom_end|_image_rodata_start|_image_rodata_end'
soc/riscv/esp32c3/linker.ld
239:    _image_rodata_start = .;
286:    _image_rodata_end = .;
``

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+endmacro()
+
+# Usage:
+#   zephyr_linker_group(NAME <name> [VMA <region>] [LMA <region>])
```
Should this be

```
 #   zephyr_linker_group(NAME <name> [VMA <region|group>] [LMA <region|group>])
```

?

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#
+# Example:
+#
+#   zephyr_linker_group(NAME groupA LMA memA)
```
I think this example would be clearer if you added

```
zephyr_linker_memory(NAME memA START ... SIZE ... FLAGS ...)
```

before this line to make this relationship between the APIs explicit.

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+# Note: <name> will be converted to lower casing for linker symbols definitions.
+#
+# NAME <name>    : Name of the group.
+# VMA <region|group>  : VMA Memory region or group to be used for this group.
```
KVMA is not documented here, and should be.

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+
+# Usage:
+#   zephyr_linker_section(NAME <name> [GROUP <group>]
+#                         [VMA <region>] [LMA <region>]
```
```
 #                         [VMA <region|group>] [LMA <region|group>]
```

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#   zephyr_linker_section(NAME <name> [GROUP <group>]
+#                         [VMA <region>] [LMA <region>]
+#                         [ADDRESS <address>] [ALIGN <alignment>]
+#                         SUBALIGN [alignment] [FLAGS <flags>]
```
```
 #                         [SUBALIGN alignment] [FLAGS <flags>]
```

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#                       If <group> is used here it means the section will use the
+#                       same LMA memory region as <group> but will not be placed
+#                       inside the group itself, see also GROUP argument.
+# GROUP <group>       : Place this section inside the group <group>
```
Can you document what is supposed to happen if I use e.g. LMA and GROUP at the same time? I'm not clear on that. Is it an error, or do I use the same LMA as the group *and* get placed inside the group?

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#                       Default a section will be present during all linker passes
+#                       But in cases a section shall only be present at a specific
+#                       pass, this argument can be used. For example to only have
+#                       this section present on the first linker pass, use `PASS 1`.
```
I think PASS can only be 1 or 2, right? Worth documenting either way and maybe worth erroring out on invalid arguments.

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#
+# Define an output section which will set up an iterable area
+# of equally-sized data structures. For use with Z_STRUCT_SECTION_ITERABLE.
+# Input sections will be sorted by name in lexicograhical order.
```
lexicographical typo

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+# Usage:
+#   zephyr_linker_section_obj_level(SECTION <section> LEVEL <level>)
+#
+# generate a symbol to mark the start of the objects array for
```
Probably worth adding a comment that this is for things like struct devices.

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+endfunction()
+
+# Usage:
+#   zephyr_linker_memory(NAME <name> START <address> SIZE <size> FLAGS <flags>)
```
The docstring makes it look like the arguments are all mandatory, but I don't see checking for missing arguments in the implementation.

That's different from `zephyr_iterable_section`, which does check if mandatory arguments are missing.

Does this need an update to improve the error handling or am I missing something?

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#   zephyr_linker_section_ifdef(<setting>
+#                               NAME <name> [VMA <region>] [LMA <region>]
+#                               [ADDRESS <address>] [ALIGN <alignment>]
+#                               SUBALIGN [alignment] [FLAGS <flags>]
```
It looks like `SUBALIGN [alignment]` has been copy/pasted around the PR a lot.

It should either be `[SUBALIGN <alignment>]` (optional argument) or `SUBALIGN <alignment>` (mandatory argument), everywhere.

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2628,3 +2662,520 @@ function(dt_chosen var)
+#
+# Each list for an input section will define the following linker symbols:
+# _<name>_list_start: Start of the iterable list
+# _<name>_list_end  : Start of the iterable list
```
end of the list

mbolivar-nordic reviewed (commented):
@tejlmand just another partial pass. I only made it up to `cmake: CMake linker functions implemented` this time again but I will take a look at the new stuff as soon as I can.

tejlmand commented:
> dev-review (Aug 19, 2021):
> ...
> 
>     * which libc are we getting?  (needs some verification, minlibc v arm-libc)

A small update on this particular item.
The arm linker will per default link to the ARM Standard C library, but that library does not contain all APIs Zephyr requires example, it only has `strlen`, not `strnlen`, secondly, the `errno.h` is more limited than what Zephyr expects.
That was the reason why minimal libc was still being linked. So some work is needed to carve out only the parts of minimal libc not provided by ARM Standard C lib, and required by Zephyr.

It is possible to disable linking with the ARM Standard library completely, by using the `-nostdlib`, `-nostdlibinc` and `-fno-built` option on the compiler, and during linking, the `--no_scanlib`.

Now, doing so all ARM C lib is disabled, including the low-level part with `__aeabi_X` implementations. (In GCC, this is provided in libgcc.a).
So it's not possible to then simply disable ARM C standard lib completely and link with Zephyr internal minimal libc or newlib, cause both of those libs relies on the low-level libgcc. (We could probably link manual to libgcc also, but then the point of using armclang kind of vanishes)

So currently we need to linking to a reduced minimal libc to be able to also link the features from Zephyr's minimal libc that is not available in ARM Standard C.

tejlmand commented:
> Now, doing so all ARM C lib is disabled, including the low-level part with `__aeabi_X` implementations. (In GCC, this is provided in libgcc.a).
> So it's not possible to then simply disable ARM C standard lib completely and link with Zephyr internal minimal libc or newlib, cause both of those libs relies on the low-level libgcc. (We could probably link manual to libgcc also, but then the point of using armclang kind of vanishes)
> 
> So currently we need to linking to a reduced minimal libc to be able to also link the features from Zephyr's minimal libc that is not available in ARM Standard C.

Please see https://github.com/zephyrproject-rtos/zephyr/commit/3a29d07c4558f9148645f459a114fbe94dc41df7 for the absolotely bare minimum at current stage.

evgeniy-paltsev commented on lib/libc/Kconfig:
```diff
@@ -11,6 +11,14 @@ config REQUIRES_FULL_LIBC
+	bool
+	default y
+
+config SUPPORT_NEWLIB_LIBC
```
Hmm, we have `TOOLCHAIN_HAS_NEWLIB` now but with this change we will also get `SUPPORT_NEWLIB_LIBC`. Do we need to differentiate them?

stephanosio replied:
Indeed. If we are going to add this, we should probably get rid of `TOOLCHAIN_HAS_NEWLIB`. Same goes for `HAS_NEWLIB_LIBC_NANO`.

IMHO, `TOOLCHAIN_HAS_NEWLIB` sounds more intuitive and in line with the existing naming convention, so maybe we could just rename it to `TOOLCHAIN_HAS_NEWLIB` and get rid of the existing CMake variable (i.e. make it fully Kconfig-derived and not reliant on the CMake scripts).

As for `SUPPORT_MINIMAL_LIBC`, `TOOLCHAIN_HAS_MINIMAL_LIBC` does not really make sense, so we could probably keep it as `TOOLCHAIN_SUPPORTS_MINIMAL_LIBC` or something like that.

tejlmand replied:
@evgeniy-paltsev  @stephanosio  Please note that the `TOOLCHAIN_HAS_NEWLIB` is a CMake variable not accessible in Kconfig.
The reason for the CMake variable can be seen here: https://github.com/zephyrproject-rtos/zephyr/pull/12393
It's intended for the test infrastructure to filter the tests based on the capabilities of the toolchain.

The `SUPPORT_NEWLIB_LIBC` is a Kconfig setting that ensures that a user cannot select `CONFIG_NEWLIB_LIBC=y` if the toolchain does not support it.
https://github.com/zephyrproject-rtos/zephyr/blob/2b91ebe16e25e5ce17270ea36b88fbb34f5f28ed/lib/libc/Kconfig#L27

Now, originally the reason for not having this is Kconfig was because @nashif wanted to avoid creating Kconfig for toolchains and instead obtain this knowledge through the CMake cache:
https://github.com/zephyrproject-rtos/zephyr/pull/12393#issuecomment-452748742
> I am not a big fan of creating a Kconfig layer for toolchains, toolchain configuration should only be cmake (toolchain file in cmake speak).

Now, that may have been true back then, but with growing number of toolchains and also toolchains coming with their own C libraries, I think we should reconsider that.

Therefore my proposal is to introduce:
- `SUPPORT_MINIMAL_LIBC`
- `SUPPORT_NEWLIB_LIBC`
  That are Kconfig symbols and can be used to ensure that only a valid C library is chosen for the toolchain.
- Consider deprecating the `TOOLCHAIN_HAS_NEWLIB` and migrate this to use the Kconfig knowledge.
   (outside the scope of this PR)
   However, before deprecating this `TOOLCHAIN_HAS_NEWLIB` we should be careful, cause we still have a desire to support using CMake toolchain files, so it could still have some value. (Using a CMake toolchain file should probably disable the libc choice menu completely)

I prefer to create new Kconfig settings for handling the C library support in Kconfig  instead of re-using an existing CMake symbol, but i'm open to another name than `SUPPORT_x_LIBC`.

Having the symbol `TOOLCHAIN_HAS_NEWLIB`, and the same symbol in Kconfig, which then becomes `CONFIG_TOOLCHAIN_HAS_NEWLIB` is bound to give confusion among readers.
Remember, all Kconfig symbols are pre-fixed with `CONFIG_<symbol>` when loaded into CMake.

stephanosio replied:
Ok, so the problem is that tests need to be filtered with `TOOLCHAIN_HAS_NEWLIB` and that cannot be done with Kconfig (at least with how the build process work right now).

> Consider deprecating the TOOLCHAIN_HAS_NEWLIB and migrate this to use the Kconfig knowledge.
> (outside the scope of this PR)

In that case, as you noted above, we should make `SUPPORT_NEWLIB_LIBC` derive from `TOOLCHAIN_HAS_NEWLIB`, especially if we are going to keep the CMake toolchain files (which I think we have no option but to do that). But, as noted in #12393, that is easier said than done due to how Zephyr builds and will likely require some rework in the build process.

That is certainly outside the scope of this PR. But then, having both `SUPPORT_NEWLIB_LIBC` in Kconfig and `TOOLCHAIN_HAS_NEWLIB` in CMake that are not synced to each other is even more confusing and error prone.

In that sense, should we defer adding these symbols until we have a more concrete solution? (maybe moved out to a separate PR?)

tejlmand replied:
> In that case, as you noted above, we should make SUPPORT_NEWLIB_LIBC derive from TOOLCHAIN_HAS_NEWLIB

This is actually not a bad idea. Let me try that.

tejlmand committed:
armclang: threading_weak.c source file added for armclang

The stub file threading_weak.c has been added containing weak stub
implementation of threading related kernel functions.

The file is needed for armlink.

When linking with armlink the linker will resolve undefined symbols for
all undefined functions even if those functions the reference the
undefined symbol is never actually called.

This file provides weak stub implementations that are compiled when
CONFIG_MULTITHREADING=n to ensure proper linking.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic reviewed (commented):
Some minor things about the latter half of the patchset.

I don't intend to review the implementations underlying the new extension API in detail. I'm going to limit myself to the higher level interface in extensions.cmake, which I think makes good sense.

The new linker scripts seem pretty experimental at this point so I agree we should not enable them by default. I also agree it doesn't make sense to keep rebasing this and keeping it out of tree. So i'm +1 on getting some basic tests passing with the new linker scripts, making sure there aren't any regressions with the existing ones, and merging as an experimental feature for LTS.

tejlmand committed:
tests: added kernel tests for arm arch with linker script generator

This commit adds an additional test case for several kernel test suites
to ensure that the linker script generator is working correctly for a
subset of the Zephyr test suites.

The ensures that the basic functionality of the linker script generator
is working while still keep the performance impact on CI at a minimal
level.

Using the kernel tests is a trade-off between testing coverage of the
linker script generator and the time it takes to complete CI.

The kernel tests is considered to have the broadest coverage of various
features important for the generated linker script.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic reviewed (approved):
My comments are addressed at this point. I talked to @tejlmand privately about the failing CI tests and am giving a +1 on the assumption those will be fixed later today.

tejlmand commented:
@nashif @carlescufi There are two failures:
1) Compliance, the libc additions for ARMClang contains typedef, just like they do for minimal, newlib, and mwdt files, so those error should be ignored.
2) The errno.h cause failure because of this:
https://github.com/zephyrproject-rtos/zephyr/blob/2bb0ed859c2e2300eaa3a8178c92b2e238e58d74/lib/libc/armstdc/include/errno.h#L8-L14
where the main part of the header originates from here:
https://github.com/zephyrproject-rtos/zephyr/blob/2bb0ed859c2e2300eaa3a8178c92b2e238e58d74/lib/libc/minimal/include/errno.h
I think we should ignore this warning, if not, please advice on what to do.

Besides that CI is green.

stephanosio reviewed (approved):
For an experimental feature, looks good enough to me.

Any nits/refinements/refactorings can be done later.
